### ISSUE TITLE
把语义分割的结果图转为灰度图发现一些像素既不属于背景也不属于自己的类别

### ISSUE THREAD

**Issue by aiwenzhu** [open-mmlab/mmdeploy#1892]
### Checklist

- [ ] I have searched related issues but cannot get the expected help.
- [ ] 2. I have read the [FAQ documentation](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/faq.md) but cannot get the expected help.
- [ ] 3. The bug has not been fixed in the latest version.

### Describe the bug

把语义分割的结果图转为灰度图发现一些像素既不属于背景也不属于自己的类别，这影响到了后边的判别

### Reproduction

把语义分割的结果图转为灰度图发现一些像素既不属于背景也不属于自己的类别，这影响到了后边的判别

### Environment

```Shell
win10 + deeplabv3P
```

**Comment by RunningLeon** [comment]
@aiwenzhu Hi, pls provide detailed info to reproduce.

**Comment by aiwenzhu** [comment]
> @aiwenzhu Hi, pls provide detailed info to reproduce.

我在项目中使用了mmsegmentation的deeplabv3p，使用了mmdeploy的部署demo:
// Copyright (c) OpenMMLab. All rights reserved.

#include "mmdeploy/segmentor.hpp"

#include <fstream>
#include <opencv2/imgcodecs/imgcodecs.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <random>
#include <string>
#include <vector>
#include <iostream>

using namespace std;
using namespace cv;

vector<cv::Vec3b> gen_palette(int num_classes) {
    std::mt19937 gen;
    //std::uniform_int_distribution<ushort> uniform_dist(0, 255);
    //std::normal_distribution<double> uniform_dist(5, 2);

    vector<cv::Vec3b> palette;
    Vec3b Color1(255,0,0);
    Vec3b Color2(0, 255, 0);
    Vec3b Color3(0, 255, 255);
    Vec3b Color4(255, 0, 255);
    Vec3b Color5(0, 128, 128);
    Vec3b Color6(128, 0, 255);
    Vec3b Color7(128,0, 128);
    Vec3b Color8(0, 0, 128);
    Vec3b Color9(128, 128, 128);
    palette.emplace_back(Color1);
    palette.emplace_back(Color2);
    palette.emplace_back(Color3);
    palette.emplace_back(Color4);
    palette.emplace_back(Color5);
    palette.emplace_back(Color6);
    palette.emplace_back(Color7);
    palette.emplace_back(Color8);
    palette.emplace_back(Color9);
    
    //palette.reserve(num_classes);
    //for (auto i = 0; i < num_classes; ++i) {
    //    palette.emplace_back(uniform_dist(gen), uniform_dist(gen), uniform_dist(gen));
    //}
    return palette;
}

int main() {
    auto device_name = "cuda";
    auto model_path = "/work_dir/trt/resnet";
    auto image_path = "/pics/test2/cut";

    vector<String> imgnames;
    glob(image_path, imgnames);

    using namespace mmdeploy;
    double t = (double)getTickCount();
    Segmentor segmentor{ Model{"/work_dir/trt/resnet"}, Device{"cuda"} };
    //Segmentor segmentor{ Model{model_path}, Device{device_name} };
    t = ((double)getTickCount() - t) / getTickFrequency();
    cout << "执行时间(秒): " << t << endl;
    for (int i = 0; i < imgnames.size(); i++)
    {
        double t = (double)getTickCount();
        cv::Mat img = cv::imread(imgnames[i]);
        int pos1 = imgnames[i].find_last_of("\\");
        string img_name = imgnames[i].substr(pos1+1,-1);
        int pos2 = img_name.find_last_of(".");
        string imgsubname = img_name.substr(0,pos2);
        string newname = string(image_path) + "/" + imgsubname + "_result.png";
        if (!img.data) {
            fprintf(stderr, "failed to load image: %s\n", image_path);
            return 1;
        }

        auto result = segmentor.Apply(img);

        auto palette = gen_palette(result->classes + 1);

        cv::Mat color_mask = cv::Mat::zeros(result->height, result->width, CV_8UC3);
        int pos = 0;
        for (auto iter = color_mask.begin<cv::Vec3b>(); iter != color_mask.end<cv::Vec3b>(); ++iter) {
            *iter = palette[result->mask[pos++]];
        }

        img = img * 0 + color_mask * 1;
        t = ((double)getTickCount() - t) / getTickFrequency();
        cout << "执行时间(秒): " << t << endl;
        cv::imwrite(newname, img);
    }
    
    return 0;
}
理论上结果图中每个像素要么是背景要么是自己分的类别，但是我将上述代码获得的彩色结果图转换成灰度图查看类别的灰度值的时候发现存在不少像素（占整张图的比例很小）的灰度值不等于背景的灰度值，也不等于自己类别的灰度值

**Comment by RunningLeon** [comment]
What's the purpose of this line `img = img * 0 + color_mask * 1;`? Doesn't this line equal  to `img= color_mask`?

**Comment by aiwenzhu** [comment]
> What's the purpose of this line `img = img * 0 + color_mask * 1;`? Doesn't this line equal to `img= color_mask`?
是的img= color_mask，本来这句是为了观察结果图在原图上覆盖位置的，img = img * 0 + color_mask * 1;

**Comment by RunningLeon** [comment]
A little confused. Did you convert `color_mask ` to gray image and compare it with original image? 
BTW,  ff the model performance is OK, then you may need to double your code.

**Comment by aiwenzhu** [comment]
> 

yes,i have compared the color_mask with the origin image,the performance is good in  training seta nd few abnormal pixels
In test sets, it's actually pretty good, but I'm also confused about the category issue. Actually, I directly use color_mask  can also see through imageJ that some pixel values are not in the 10 colors which in the palette in my code above. Most of these abnormal pixels are located at the edge of a category area

**Comment by aiwenzhu** [comment]
> > 
> 
> yes,i have compared the color_mask with the origin image,the performance is good in training seta nd few abnormal pixels In test sets, it's actually  good, but I'm also confused about the category issue. Actually, I directly use color_mask can also see through imageJ that some pixel values are not in the 10 colors which in the palette in my code above. Most of these abnormal pixels are located at the edge of a category area
